Python/eof.py

```diff
@@ -0,0 +1,198 @@
+import numpy as np
+from sklearn.decomposition import pca 
+from sklearn import linear_model
+
+from operators import *
+
+def ensemble_reshape(data_cube):
+    """Reshapes data cube into array shape 
+    for eof calculations. 
+    --------------------------------
+    Args: 
+        data_cube (xarray Dataset): datacube 
+            containing an ensemble of 
+            CanRCM4 models 
+    Returns:
+        field_ens (numpy.ndarray): reshaped 2-d array
+            into number of cells x ensemble size
+    """
+    check_keys(data_cube)
+
+    x_field = data_cube['dv']
+    n, p = cell_count(data_cube)
+    field_ens = np.reshape(x_field.values, (p, n))
+
+    return field_ens
+
+def get_land(ensemble_arr):
```

Comment:
Is a `numpy.ndarray` containing indices of either land or sea acceptable?

Reply:
Well, most GCM's come with land/sea masks. And I *think* (this is maybe hypothetical) when you read them in you should get a [`numpy maskedarray`](https://docs.scipy.org/doc/numpy/reference/maskedarray.generic.html). There's pretty standard canonical ways to [get the indices from the mask](https://docs.scipy.org/doc/numpy/reference/maskedarray.generic.html#accessing-only-the-valid-entries) that are short enough that it might be worthwhile to simply use them inline.

I don't know, it's a judgement call, but I tend to think that it's better to pass around something that already has a defined interface and a library of software built around it, then to define something yourself. As long as the already-written-thing is not over-the-top heavyweight (which numpy maskedarrays are not).

Reply:
Would it be acceptable to leave the masking to the user, but then provide a useful function like `mask_nan` which [masks a data cube with xarray?](http://xarray.pydata.org/en/stable/indexing.html#masking-with-where) There is just a lot extra code I'd need to add to check and apply a mask in every case.

Reply:
i.e. the `data_cube` provided is already masked.

Reply:
Yeah, I guess if the `data_cube` is already masked (or if we can assume or check that it will be), then maybe this is a case where the code itself doesn't need to change that much, but the way that you describe it does? Like, do we *actually* care whether it's land or sea, or do we just care whether there is data or its masked?

Reply:
I've spent a lot of time thinking about this, and playing around a bit. The process to get the `data_cube` into a state where we can run EOF on it, is something like this: `xarray data_cube` -> `np.ndarray ensemble_array (shape grid cells x ensemble size)` -> `eof`.

Previously, I was applying the mask like this:  `xarray data_cube` -> `np.ndarray ensemble_array (shape grid cells x ensemble size)` -> `MASK NAN (shape of ensemble_array)` -> `eof`.

I think it's unnecessarily complicated to mask using xarray (and it doesn't really work better than `numpy.ma`), and its far less intuitive. So instead, I think I will do something like this:

`xarray data_cube` -> `MASK` -> `reshape MASK to ensemble_array shape` -> `reshape data_cube to np.ndarray ensemble_array` -> `eof`.

This reshaping that's necessary makes it a little tricky. I think the most logical thing to do is to mask the values in its original `CanRCM4` shape - then do _any_ transformations I do to the data, to the mask so that I can use the mask on the transformed shape. Does that make sense?

Reply:
I'm going to elaborate a little bit more.

It's hard to remove values from a grid with spatial information, and select only the data values of interest _without first_ transforming the data. 

What shape nd.array would be returned if I mask ocean and only got land? How can one remove `NaN` values from a grid, and still maintain its original shape? Those grid cells are important because they contain the location information of a cell, and you can't have asymmetrically shaped `ndarrays` as far as I know.

What makes more sense, is to unravel the grids so that each grid cell across an _ensemble_ is found as a row in the `ensemble arr`. An ocean grid cell (which in CanRCM4 grid contains `NaN`) will be `NaN` for _each_ ensemble, and thus will be a row of `NaN` values in the ensemble array. This way, it's easier to mask the `NaN` rows, and only use the data values.

Thus my thinking that masking _can_ occur in the CanRCM4 grid shape, but _must_ be transformed into an `ensemble arr` shape before data can be extracted from that mask for eof analysis. So we have a choice, either construct a mask that makes sense in the CanRCM4 grid shape and then is transformed, or construct a mask from the get-go that requires less transformation since its already in the `ensemble arr` shape.